### ISSUE TITLE
Three questions about the code and paper

### ISSUE THREAD

**Issue by rkotimi** [hlsheng1/CT3D#10]
Hello @hlsheng1 ! After reading your paper and code, I want to ask some questions.

1. Why did you transform the coordinates of points from cartesian coordinate system to spherical coordinate system in proposal-to-point embedding module? Did experiments show better performance in spherical coordinate system than in cartesian coordinate system?
2. CT3D is trained with more epochs and a smaller learning rate compared to the vanilla SECOND in OpenPCDet. Is this because Transformer converges slower than other networks?
3. I can't fully understand the proposed extended channel-wise re-weighting. In my opinion, it's more like head-wise re-weighting for the shape of **scores_1** is (batch_size * num_roi, num_head, num_query, num_key). Besides, extended channel-wise re-weighting doesn't introduce more learnable parameters. Why would the performance be better?

I am looking forward to your reply. Thank you in advance!

**Comment by hlsheng1** [comment]
@rkotimi Thank you for your attention.
1. The spherical coordinate has minor effect on the final performance, and that is the reason I don't put it in the paper. In my experiments, the performance of removal or not is almost the same. You can have a try by yourself.
2. Yes, most transformer based network adopt the strategy of pre-training, our network is trained form the scratch. I suggest using a smaller learning rate.
3. I recommand that you can read the paper instead of the simplified code. The output dimension of the standard decoder is one aggregated feature (e.g. 1 * 256(feature num)), however, our extend channel-wise re-weighting outputs a feature map (e.g. 1 * 256(feature num) * 256(channel num)), so we add one extra linear function to aggregate the features (mathmatically, it is equal to aggregate the attention weights). So, I have two viewpoint on your question: Firstly, the number of learnable parameters could not represent the model ability, instead, a more elegent and efficient model design is needed (a simpel example: dose all the two-stage detectors better than the single-stage detectors?) Secondly, we indeed add some learnable parameters, i.e., the line 223 of ctrans.py.

If you have more questions, please let me know. We can have more detailed discusses.

**Comment by rkotimi** [comment]
Thank you for your detailed reply! In fact I have read your paper several times, but it is hard for me to totally understand the channel-wise decoding module and associate it with the code implementation. Now I still have some questions.

1. What do the two items in the blue and red bounding box represent, respectively?
![Picture1](https://user-images.githubusercontent.com/61943053/146196766-6ef9b8c6-aaee-4add-aea3-edd080a83926.png)
2. How to modify the **attention** function to implement channel-wise re-weighting, instead of extended channel-wise re-weighting?
https://github.com/hlsheng1/CT3D/blob/759142f112b07c4f25579466d6ed8effb2ec5b26/pcdet/models/model_utils/ctrans.py#L206
3. The third question is about the implementation details of Transformer. In the self attention module of **TransformerDecoderLayer**,  the query and key are equal and learnable parameters, but the elements of the value tensor is a lot of zeros.  So far as I know, query, key and value should be the same in the self attention module. Could you explain why the value is unequal to query and key? 
https://github.com/hlsheng1/CT3D/blob/759142f112b07c4f25579466d6ed8effb2ec5b26/pcdet/models/model_utils/ctrans.py#L272

I think CT3D is an excellent work and it is worth taking some time to explore it. I hope you can help me better understand it. Thank you once again!

**Comment by hlsheng1** [comment]
@rkotimi I am glad to address your questions.
1. Your red box is the final decoding weights. The dimension after the endoder is not changed, so we have to aggregate the encoded point features into one global representation. So, every point needs one decoding weights, and that is how the decoder works. Your blue box is the introduced weight. You can put it together with the follwing sum operator, and this is actually the matrix multiplication between one matrix and one vector.
2. The implementation is simple. You can do softmax(key, dim=-1) at line 210 and drop out the score_1/score_2.
3. I think your conclusion is not right to my best knowledge. No matter the self-attention or the (cross-)attention mechanisms, query, key and value should not be the same. In self-attention mechanism, these three come from same input features, but then are processed by different linear layers. In decoder (cross-)attention mechanism, only query is learnable parameters, key is still comes the input linear projection. Mathmatically, self-attention is like to do a Autocorrelation (cov(X,X)), cross-attention in decoder is like to do crosscorrelation (cov(X,Y)). Clearly, if key is also totally the learnable parameters, there is no sense to make a correlation computation, so as to get the final attention weights.
3+. Your mentioned "Lots of zeros" is right. I also found this phenomenon at the end of the training stage. I think it is hard to explain the data generated in the middle of the neural network.

**Comment by rkotimi** [comment]
Thank you for your detailed reply! I have no more questions about the channel-wise re-weighting module. For question 3, maybe my expression was not rigorous enough. The query, key and value I mentioned above are actually the input feature before being processed with  linear projection. So I argue that them should the same. 

In the forward function of Transformer, the variable **tgt** is created as a tensor filled with zeros.
https://github.com/hlsheng1/CT3D/blob/759142f112b07c4f25579466d6ed8effb2ec5b26/pcdet/models/model_utils/ctrans.py#L74
 The **tgt** is passed into self.decoder and added with learnable parameter **query_pos**, so as to get **q** and **k**.
https://github.com/hlsheng1/CT3D/blob/759142f112b07c4f25579466d6ed8effb2ec5b26/pcdet/models/model_utils/ctrans.py#L271
So query and key are from the same input feature, namely **query_pos**. However, value is equal to **tgt**, which is filled with zeros. That means "Lots of zeros" has nothing to do with network training.

I also reference the pytorch official version of TransformerDecoderLayer. In their implementation, the three inputs to self attention mechanism module are exactly the same.
https://github.com/pytorch/pytorch/blob/5cc4037369bd9a91b6c1bc1b6ff5a8dca91d015e/torch/nn/modules/transformer.py#L462

Pardon me if it is a misunderstanding caused by my poor knowledge.

**Comment by hlsheng1** [comment]
@rkotimi Hi, Sry I thought you were referring the cross-attention part of the decoder. The decoder consists of two parts: one is self-attention based on the learnable query, another is cross-attention based on the learnable  query and the input. 

The codes at line 271-275 actually perform self-attention based on the learnable query. However, we only have only one query, and the initial target is zero, so "line 271-275" can be just dropped out in this implementation. These codes indeed have nothing to do with network traininig, and I chose to keep them because of possible extensions, such as multiple queries or multi-layer decoders.
